particle-iot/device-os
# [Gen 3] 256KB application support

### Description

Refer to the [SDD](https://docs.google.com/document/d/1DvOFlKDMDqB1JotDDHTFnWLNPyqYZ9ZADfSXr4W8P9U/edit)

NOTE: this PR also enables LTO on all Gen 3 platforms when building system parts.

**NOTE: do not use current version of CLI to flash new 256KB binaries over DFU, this will fail**

### Steps to Test

NOTE: `3.1.0-alpha.1` and `3.1.0-alpha.2` releases available on staging can be used for testing (`test/v3.1.0-alpha.1` and `test/v3.1.0-alpha.2` are the respective branches). The releases are fully available on staging and appropriate version of DS is also deployed there.

1. Building larger than 128KB applications should be possible with this branch. They should also run just fine
2. 128KB older applications should still work correctly
3. `particle serial inspect` should correctly report either type of an application (index = 1, 128KB max) (index = 2, 256KB max)
4. Flashing 128KB older app should override the current 256KB app, even if they don't overlap
5. Flashing 256KB app should override (invalidate) the older 128KB  app residing on the device even if they don't overlap
6. Bootloader should function correctly with older versions of system-part1 on the device (mainly the bootloader updates through MBR are a concern and application updates)

### Example App

N/A

### References

- [CH60338]
- There is an appropriate PR in DS repo

---

### Completeness
